### ISSUE TITLE
Pointers to Entities?

### ISSUE THREAD

**Issue by hiptmair** [craffael/lehrfempp#32]
I am planning  to devise a data structure for a hierarchy of meshes  generated by (loca) refinement. This data structure should make possible accessing parents and children  of mesh entities, _if they exist_.

Is storing pointers `const Entity *` to parent and child entities compatible with the current design of LehrFEM++? The issue  is also how to obtain these pointers, because the creation of Entity objects is now buried deep in the constructor of a `Mesh`, which is used by a `MeshFactory`, which should be employed by the refinement algorithm. 

I cannot even use indices, because the acess index -> entity object is not supported!

Basically, the challenge is to endow each entity with information about a potential parent entity, maybe in the form of an array  of `const Entity *`. I have  no idea how to build such an array, however.

**Comment by hiptmair** [comment]
A solution is that the methods `AddPoint()` and `AddEntity()` of a `MeshFactory` reliably return the final index of an entity in the mesh. This seems to be the case in the current implementation in hybrid2dp, but remains to be checked.

**Comment by craffael** [comment]
1) Yes, This should actually be the case! If you check the [doxygen documentation of the MeshFactory interface](https://craffael.github.io/lehrfempp/classlf_1_1mesh_1_1_mesh_factory.html) it says that the `AddPoint()` and `AddEntity()` methods return the index of the created entity. Maybe this could be clarified even more by adding a reference to the `lf::mesh::Mesh::Index()` method.

2) In principle you can store `const Entity*` pointer for the father/child entities. However, the mesh returns you only `ForwardIterator<Entity>`'s. Since the latter behave almost exactly like a pointer, you could also store them.

3) Yes, with the current design it is impossible to retrieve an entity with a given index in constant time. Of course we could add another Method to the `lf::mesh::Mesh` interface class that allows us to retrieve an entity based on its index. However, I think in this case we can solve the problem elegantly by storing `ForwardIterator<Entity>` objects that point to the father/child objects.

4) In the Gmsh-Branch I have already introduced the datastructure `lf::mesh::GridDataSet` which can be used to store information with entities (such as pointers to father/child entities). I had to introduce such a structure in order to store the Gmsh Physical tags with every entity but I think it will also be used to store flags/material parameters with entities in the future.
If you want you can merge the gmsh branch into your development branch and then you can start to use this structure as well (e.g. you could store the pointers to father entities of one mesh in a   std::shared_ptr<MeshDataSet<ForwardIterator<Entity>>>> object). At the moment everything should compile in the gmsh-branch, but the gmsh reader is not yet fully tested/documented.

**Comment by craffael** [comment]
Yeah I know, this is still a short-coming that currently ForwardIterators don't inherit from a generic "pointer" type that only allow dereferencing, maybe I should implement this soon.
